# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/sandbox/chromium/sandbox/win/src/integrity_level_test.cc
# Commit: 370c40e49230
# Full Hash: 370c40e4923073a87845b57c1f4851bacd9663da
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-12 21:42:35
# Regressor Bug: 1552160
# File Overlap Count: 3
# Description:
#   Bug 1552160 Part 1: Roll-up of chromium sandbox update and mozilla patches to get a running browser. r=jld,aklotz,tjr,bobowen
#   
#   This updates security/sandbox/chromium/ files to chromium commit 84108231f6e6e0772fb9a4643679ce76aa771e67.
#   
#   Existing and new patches applied from security/sandbox/chromium-shim/patches/with_update/ to give a compiling and mostly working browser.
# ==============================================================================

diff -r bd5ecd72f76c -r 370c40e49230 security/sandbox/chromium/sandbox/win/src/integrity_level_test.cc
--- a/security/sandbox/chromium/sandbox/win/src/integrity_level_test.cc	Wed Jun 12 21:55:59 2019 +1200
+++ b/security/sandbox/chromium/sandbox/win/src/integrity_level_test.cc	Wed Jun 12 11:10:48 2019 +0100
@@ -3,19 +3,32 @@
 // found in the LICENSE file.
 
 #include <windows.h>
+
 #include <atlsecurity.h>
 
+#include "base/process/process_info.h"
 #include "base/win/windows_version.h"
-#include "testing/gtest/include/gtest/gtest.h"
 #include "sandbox/win/src/sandbox.h"
+#include "sandbox/win/src/sandbox_factory.h"
 #include "sandbox/win/src/sandbox_policy.h"
-#include "sandbox/win/src/sandbox_factory.h"
 #include "sandbox/win/tests/common/controller.h"
+#include "testing/gtest/include/gtest/gtest.h"
 
 namespace sandbox {
 
+SBOX_TESTS_COMMAND int CheckUntrustedIntegrityLevel(int argc, wchar_t** argv) {
+  return (base::GetCurrentProcessIntegrityLevel() == base::UNTRUSTED_INTEGRITY)
+             ? SBOX_TEST_SUCCEEDED
+             : SBOX_TEST_FAILED;
+}
 
-SBOX_TESTS_COMMAND int CheckIntegrityLevel(int argc, wchar_t **argv) {
+SBOX_TESTS_COMMAND int CheckLowIntegrityLevel(int argc, wchar_t** argv) {
+  return (base::GetCurrentProcessIntegrityLevel() == base::LOW_INTEGRITY)
+             ? SBOX_TEST_SUCCEEDED
+             : SBOX_TEST_FAILED;
+}
+
+SBOX_TESTS_COMMAND int CheckIntegrityLevel(int argc, wchar_t** argv) {
   ATL::CAccessToken token;
   if (!token.GetEffectiveToken(TOKEN_READ))
     return SBOX_TEST_FAILED;
@@ -30,11 +43,11 @@
   TOKEN_MANDATORY_LABEL* label =
       reinterpret_cast<TOKEN_MANDATORY_LABEL*>(buffer);
 
-  PSID sid_low = NULL;
+  PSID sid_low = nullptr;
   if (!::ConvertStringSidToSid(L"S-1-16-4096", &sid_low))
     return SBOX_TEST_FAILED;
 
-  BOOL is_low_sid = ::EqualSid(label->Label.Sid, sid_low);
+  bool is_low_sid = ::EqualSid(label->Label.Sid, sid_low);
 
   ::LocalFree(sid_low);
 
@@ -45,7 +58,6 @@
 }
 
 TEST(IntegrityLevelTest, TestLowILReal) {
-
   TestRunner runner(JOB_LOCKDOWN, USER_INTERACTIVE, USER_INTERACTIVE);
 
   runner.SetTimeout(INFINITE);
@@ -60,7 +72,6 @@
 }
 
 TEST(DelayedIntegrityLevelTest, TestLowILDelayed) {
-
   TestRunner runner(JOB_LOCKDOWN, USER_INTERACTIVE, USER_INTERACTIVE);
 
   runner.SetTimeout(INFINITE);
@@ -74,7 +85,6 @@
 }
 
 TEST(IntegrityLevelTest, TestNoILChange) {
-
   TestRunner runner(JOB_LOCKDOWN, USER_INTERACTIVE, USER_INTERACTIVE);
 
   runner.SetTimeout(INFINITE);
@@ -82,4 +92,27 @@
   EXPECT_EQ(SBOX_TEST_DENIED, runner.RunTest(L"CheckIntegrityLevel"));
 }
 
+TEST(IntegrityLevelTest, TestUntrustedIL) {
+  TestRunner runner(JOB_LOCKDOWN, USER_RESTRICTED_SAME_ACCESS, USER_LOCKDOWN);
+  runner.GetPolicy()->SetIntegrityLevel(INTEGRITY_LEVEL_LOW);
+  runner.GetPolicy()->SetDelayedIntegrityLevel(INTEGRITY_LEVEL_UNTRUSTED);
+  runner.GetPolicy()->SetLockdownDefaultDacl();
+
+  runner.SetTimeout(INFINITE);
+
+  EXPECT_EQ(SBOX_TEST_SUCCEEDED,
+            runner.RunTest(L"CheckUntrustedIntegrityLevel"));
+}
+
+TEST(IntegrityLevelTest, TestLowIL) {
+  TestRunner runner(JOB_LOCKDOWN, USER_RESTRICTED_SAME_ACCESS, USER_LOCKDOWN);
+  runner.GetPolicy()->SetIntegrityLevel(INTEGRITY_LEVEL_LOW);
+  runner.GetPolicy()->SetDelayedIntegrityLevel(INTEGRITY_LEVEL_LOW);
+  runner.GetPolicy()->SetLockdownDefaultDacl();
+
+  runner.SetTimeout(INFINITE);
+
+  EXPECT_EQ(SBOX_TEST_SUCCEEDED, runner.RunTest(L"CheckLowIntegrityLevel"));
+}
+
 }  // namespace sandbox